scripts/bulkimport.sh

```diff
@@ -0,0 +1,240 @@
+		OPTARG="${OPTARG#$OPT}"  # extract long option argument (may be empty)
+		OPTARG="${OPTARG#=}"     # if long option argument, remove assigning `=`
+	fi
+	# shellcheck disable=SC2214
+	case "${OPT}" in
+	u | users)
+		resources+=(users)
+		;;
+	r | repos)
+		resources+=(repos)
+		;;
+	g | groups)
+		resources+=(groups)
+		;;
+	a | artifactory-url)
+		host="${OPTARG}"
+		grep -qE 'http(s)?://.*' <<<"${host}" || (echo "malformed url name: ${host}. must be of the form http(s)?://.*" && exit 1)
+		;;
+	??*)
+		usage
+		;;
+	*)
+		usage
+		;;
+	esac
+done
+
+function netrc_location {
+		case "$(uname)" in
+  		CYGWIN* | MSYS* | MINGW*)
+  			echo "$HOME/_netrc"
+  			;;
+  		Darwin* | Linux*)
+  			echo "$HOME/.netrc"
+  			;;
+  		*)
+  			echo "unsupported OS" >&2
+  			return 255
+  			;;
+  	esac
+}
+function assert_netrc {
+	local location
+	location="$(netrc_location)"
+	test -f "${location}" || touch "${location}"
+	echo "${location}"
+}
+
+function hasNetRcEntry {
+	local h="${1:?No host supplied}"
+
+	h="${h/https:\/\//}"
+  h="${h/http:\/\//}"
+  grep -qE "machine[ ]+${h}" "$(assert_netrc)"
+}
+
+function write_netrc {
+	local host="${1:?You must supply a host name}"
+	local netrc
+	netrc=$(assert_netrc)
+	read -r -p "please enter the username for ${host}: " username
+	read -rs -p "enter the api token (will not be echoed): " token
+	# append only to both files
+	cat <<-EOF  >> "${netrc}"
+		machine ${host}
+		login ${username}
+		password ${token}
+	EOF
+	echo "${host}"
+}
+
+# if they have no netrc file at all, create the file and add an entry
+if ! hasNetRcEntry "${host}" ; then
+	echo "added entry to $(write_netrc "${host}" )  needed for curl" >&2
+fi
+
+function repos {
+	# jq '.resources |map({type,name})' terraform.tfstate # make sure to not include anything already in state
+	# we'll make our internal jq structure match that of the tf state file so we can subtract them easy
+	local host="${1:?You must supply the artifactory host}"
+	# literally, usage of jq is 6x faster than bash/read/etc
+	# GET "${host}/artifactory/api/repositories" returns {key,type,packageType,url} where
+	# url) points to the UI for that resource??
+	# packageType) is cased ??
+	# type) is upcased and in "${host}/artifactory/api/repositories/${key}" it's not AND it's called rclass
+	local url="${host}/artifactory/api/repositories"
+	curl -snLf "${url}" | jq -re  --arg u "${url}" '.[] | "\($u)/\(.key)"' |
+		xargs -P 10 curl -snLf |
+			jq -sre '
+				group_by(.packageType == "docker" and .rclass == "local") |
+				(.[0] | map({
+						type: "artifactory_\(.rclass)_\(.packageType)_repository.\(.key)",
+						name: key
+					})
+				) +
+				(.[1] | map({
+						type: "artifactory_\(.rclass | ascii_downcase)_\(.packageType | ascii_downcase)_\(.dockerApiVersion | ascii_downcase)_repository.\(.key)",
+						name: key
+					})
+				) | .[] |
+"import {
+  to = \(.type)
+  id = \"\(.name)\"
+}"'
+} && export -f repos
```

Comment:
compare the above function for repos with the current one that makes a carve out for docker-local and _only_ fetches those:


```shell

function repos {
	# jq '.resources |map({type,name})' terraform.tfstate # make sure to not include anything already in state
	# we'll make our internal jq structure match that of the tf state file so we can subtract them easy
	local host="${1:?You must supply the artifactory host}"
	# literally, usage of jq is 6x faster than bash/read/etc
	# GET "${host}/artifactory/api/repositories" returns {key,type,packageType,url} where
	# url) points to the UI for that resource??
	# packageType) is cased ??
	# type) is upcased and in "${host}/artifactory/api/repositories/${key}" it's not AND it's called rclass
	local tempJson
	tempJson=$(mktemp)
	local url="${host}/artifactory/api/repositories"
	# we have to sort out the wheat from the chaffee. We're normalizing the input while we sort it out
	# and we choose to map to 'rclass' from '.type' because that's how the Go code maps it
	curl -snLf "${url}" |
		jq 'map({key, rclass: (.type | ascii_downcase), packageType : (.packageType | ascii_downcase)}) |
				group_by(.packageType == "docker" and .rclass == "local") |
				{
					safe: .[0],
					docker_remap: .[1]
				}
		' > "${tempJson}"
  # the URL that comes in the original payload refers to the UI endpoint. Dumb
	jq -re  --arg u "${url}" '.docker_remap[] | "\($u)/\(.key)"' "${tempJson}" |
		#grab the docker-local repos
		xargs -n 10 -P 10 curl -snLf |
		# this was literally the only field we couldn't get from before and, apparently it's no longer possible to
		# even set docker V1 in RT (even though there is a check, you get an error if you try). But for legacy reason, we
		# have to go fetch them. This would be 1 line to simply fetch all repo data and remap it. But SOMEONE is worried about
		# scalability
		jq -sre 'map(.dockerApiVersion |= ascii_downcase)' |
			# combined step 1 with the tf state and step 3, and give them saner names
			cat groups.json terraform.tfstate - | jq -sre '
				{
					safe: .[0].safe,
					tf: .[1].resources,
					docker:.[2:][0]
				} |
				((.safe | map({
						type: "artifactory_\(.rclass)_\(.packageType)_repository.\(.key | ascii_downcase)",
						name: .key
					})
				) +
				(.docker | map({
						type: "artifactory_\(.rclass)_\(.packageType)_\(.dockerApiVersion)_repository.\(.key)",
						name: .key
					})
				)) -
				(.tf | map({
						type: "\(.type).\(.name)",
						name: .name
					})
        ) '
} && export -f repos
```

This code also actually compares to the local tfstate file and removes anything that's already being tracked. Not sure how we feel about it